orderer/common/cluster/clusterservice.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright IBM Corp. 2017 All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package cluster
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"strconv"
+	"time"
+
+	"github.com/hyperledger/fabric-protos-go/orderer"
+	"github.com/hyperledger/fabric/common/flogging"
+	"github.com/hyperledger/fabric/common/util"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+//go:generate mockery -dir . -name Dispatcher -case underscore -output ./mocks/
+
+// Dispatcher dispatches requests
+/* type ClusterDispatcher interface {
+	DispatchSubmit(ctx context.Context, request *orderer.NodeTransactionOrderRequest) error
+	DispatchConsensus(ctx context.Context, request *orderer.NodeConsensusRequest) error
+}
+*/
+//go:generate mockery -dir . -name StepStream -case underscore -output ./mocks/
+
+// StepStream defines the gRPC stream for sending
+// transactions, and receiving corresponding responses
+type ClusterStepStream interface {
+	Send(response *orderer.ClusterNodeServiceStepResponse) error
+	Recv() (*orderer.ClusterNodeServiceStepRequest, error)
+	grpc.ServerStream
+}
+
+// Service defines the raft Service
+type ClusterService struct {
+	StreamCountReporter              *StreamCountReporter
+	Dispatcher                       Dispatcher
+	Logger                           *flogging.FabricLogger
+	StepLogger                       *flogging.FabricLogger
+	MinimumExpirationWarningInterval time.Duration
+	CertExpWarningThreshold          time.Duration
+}
+
+// Step passes an implementation-specific message to another cluster member.
+func (s *ClusterService) Step(stream orderer.ClusterNodeService_StepServer) error {
+	s.StreamCountReporter.Increment()
+	defer s.StreamCountReporter.Decrement()
+
+	addr := util.ExtractRemoteAddress(stream.Context())
+	commonName := commonNameFromContext(stream.Context())
+	exp := s.initializeExpirationCheck(stream, addr, commonName)
+	s.Logger.Debugf("Connection from %s(%s)", commonName, addr)
+	defer s.Logger.Debugf("Closing connection from %s(%s)", commonName, addr)
+
+	var channel string
+	var err error
+	for {
+		channel, err = s.handleMessage(stream, addr, exp, channel)
```

Comment:
the sequence/order of message is handled within handleMessage. Condition at line no. 95 validates & returns error if it out of order.